add special behavior of setPrototypeOf (#23018)
mdn/content: files/en-us/web/javascript/reference/global_objects/object/setprototypeof/index.md

- {{jsxref("TypeError")}}
  - : Thrown if one of the following conditions is met:

>     - The `obj` parameter is [non-extensible](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/isExtensible), or it's an [immutable prototype exotic object](https://tc39.es/ecma262/#sec-immutable-prototype-exotic-objects), such as `Object.prototype` or [`window`](/en-US/docs/Web/API/Window), but the engine will check if you passed the exact current `obj` prototype as a second parameter no errors will be thrown,
>     and The function will return `obj` as it is.

Reviewer: ```suggestion
    - The `obj` parameter is [non-extensible](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/isExtensible), or it's an [immutable prototype exotic object](https://tc39.es/ecma262/#sec-immutable-prototype-exotic-objects), such as `Object.prototype` or [`window`](/en-US/docs/Web/API/Window), and the new prototype is not the same value as the original prototype of `obj`.
```

Author: This is simpler for sure, I just don't want the reader to think that this condition is linked with exotic objects only, how about 
  "and for both cases, the new prototype is not the same value as the original prototype of `obj`"

Reviewer: Then what about:

```suggestion
    - The `obj` parameter is [non-extensible](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/isExtensible), or it's an [immutable prototype exotic object](https://tc39.es/ecma262/#sec-immutable-prototype-exotic-objects), such as `Object.prototype` or [`window`](/en-US/docs/Web/API/Window). However, the error is not thrown if the new prototype is the same value as the original prototype of `obj`.
```

Author: Perfect👏

```js
Object.isExtensible(Object.prototype); // true; you can add more properties
Object.setPrototypeOf(Object.prototype, {}); // TypeError: Immutable prototype object '#<Object>' cannot have their prototype set

> Object.setPrototypeOf(Object.prototype, Object.getPrototypeOf(Object.prototype)) // No error; the prototype chain is not modified

Reviewer: Oh also, `Object.getPrototypeOf(Object.prototype)` is just `null`. I'm suggesting you replace it, because otherwise this line gets really long and I want to make it shorter with all efforts.

```suggestion
Object.setPrototypeOf(Object.prototype, null); // No error; the prototype chain is not modified
```

Author: You are going the opposite direction 😄 , this makes it a little bit harder to reason about without a comment

Reviewer: "the prototype chain is not modified" is a pretty self-evident description. It both explains the code itself and the result.

Author: what about making a variable carrying the prototype and passing it

Reviewer: Okay but that's still literally just `null`. Maybe changing the sentence would help, if you insist people won't understand:

```suggestion
Object.setPrototypeOf(Object.prototype, null); // No error; the prototype of `Object.prototype` is already `null`
```

Author: Thank you, Sida ❤️, and it's up to you, just do what you think is better, I'm just trying to discuss with you what's better for people to get the example.

Author: > 

I think the last suggestion is much better for both, so let's stick with it, unless you think the prev is better

Author: > Okay but that's still literally just `null`. Maybe changing the sentence would help, if you insist people won't understand:

The idea of making a variable is just to let the reader know where does the value = I mean `null` come from

```js
Object.isExtensible(Object.prototype); // true; you can add more properties
Object.setPrototypeOf(Object.prototype, {}); // TypeError: Immutable prototype object '#<Object>' cannot have their prototype set

> Object.setPrototypeOf(Object.prototype, Object.getPrototypeOf(Object.prototype)) // No errors as you pass the same current prototype of Object.prototype, and Object.prototype will be returned as it is.

Reviewer: ```suggestion
Object.setPrototypeOf(Object.prototype, Object.getPrototypeOf(Object.prototype)) // No error; the prototype chain is not modified
```

Author: How about "// No error; the prototype chain is not modified because you passed the same original prototype as a new prototype"

Author: I just think mentioning the reason will be better in this context

Reviewer: I also gave the suggestion above to add this sentence to the description prose ^

Author: I think adding it here will be better as well, but Let's stick with yours. 
Thank you.

Reviewer: Oh also, `Object.getPrototypeOf(Object.prototype)` is just `null`. I'm suggesting you replace it, because otherwise this line gets really long and I want to make it shorter with all efforts.
